packages/utilities/src/classNamesFunction.ts

```diff
@@ -12,29 +25,114 @@ export interface IClassNamesFunctionOptions {
-    // `props.styles` itself is defined as optional, this avoids the need to use `!` at all invocation points.
-    if (styleFunctionOrObject === undefined) {
-      return {} as IProcessedStyleSet<TStyleSet>;
+    let current: Map<any, any> = map;
+    const disableCaching = options.disableCaching;
+
+    if (!options.disableCaching) {
+      current = _traverseMap(map, styleFunctionOrObject as any);
+      current = _traverseMap(current, styleProps);
     }
 
-    const styleSet =
-      styleFunctionOrObject && (typeof styleFunctionOrObject === 'function' ? styleFunctionOrObject(styleProps!) : styleFunctionOrObject);
+    if (disableCaching || !(current as any)[RetVal]) {
+      if (styleFunctionOrObject === undefined) {
+        (current as any)[RetVal] = {} as IProcessedStyleSet<TStyleSet>;
+      } else if (Array.isArray(styleFunctionOrObject)) {
+        (current as any)[RetVal] = mergeStyleSets(...(styleFunctionOrObject as any).map(_derive, styleProps));
+      } else {
+        (current as any)[RetVal] = mergeStyleSets(
+          typeof styleFunctionOrObject === 'function' ? styleFunctionOrObject(styleProps) : styleFunctionOrObject
+        );
+      }
+
+      if (!disableCaching) {
+        resultCount++;
+      }
+    }
+
+    if (resultCount > MAX_CACHE_COUNT) {
+      map.clear();
+      resultCount = 0;
+
+      // Mutate the options passed in, that's all we can do.
+      options.disableCaching = true;
+
+      // Note: this code is great for debugging problems with styles being recaculated, but commenting it out
+      // to avoid confusing consumers.
+
+      // if (process.env.NODE_ENV !== 'production') {
+      //  console.log('Styles are being recalculated far too frequently. Something is mutating the class over and over.');
+      //  // tslint:disable-next-line:no-console
+      //  console.trace();
+      // }
+    }
 
-    return mergeStyleSets(styleSet as TStyleSet);
+    // Note: the RetVal is an attached property on the Map; not a key in the Map. We use this attached property to
+    // cache the return value for this branch of the graph.
+    return (current as any)[RetVal];
   };
 
   return getClassNames;
 }
+
+function _derive(obj: any): any {
+  return typeof obj === 'function' ? obj(this) : obj;
+}
+
+function _traverseEdge(current: Map<any, any>, value: any): Map<any, any> {
+  value = _normalizeValue(value);
+
+  if (!current.has(value)) {
+    current.set(value, new Map<any, any>());
+  }
+
+  return current.get(value);
+}
+
+function _traverseMap(current: Map<any, any>, inputs: any[] | Object): Map<any, any> {
+  if (Array.isArray(inputs)) {
+    for (let i = 0; i < inputs.length; i++) {
+      current = _traverseEdge(current, inputs[i]);
+    }
+  } else if (typeof inputs === 'object') {
+    for (const propName in inputs) {
+      if (inputs.hasOwnProperty(propName)) {
+        current = _traverseEdge(current, (inputs as any)[propName]);
+      }
+    }
```

Comment:
Should there be some kind of error thrown if a primitive value somehow gets passed into inputs?

Reply:
Currently this is limited by typing via classNamesFunction. We generally don't throw errors for cases that are limited through typing. I do agree, however, that typing should be improved, but since `_traverseMap` isn't exported, I think this is ok for now.